pentaho-proc-governor/impl/src/main/java/org/hitachivantara/process/ProcessGovernorImpl.java

```diff
@@ -0,0 +1,134 @@
+/*!
+ * Copyright 2010 - 2018 Hitachi Vantara.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.hitachivantara.process;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Semaphore;
+
+import static java.lang.String.format;
+
+/**
+ * A ProcessGovernor implementation which launches separate threads for each proc execution request,
+ * used for managing proc execution and completion.
+ *
+ *
+ * Current limitations
+ * 1)  Use any sort of defined queueing.  It's arbitrary which thread will be able to launch its proc when
+ * a permit becomes available.
+ * 2)  Support running procs under a different account.  Everything run under the permissions granted to the user
+ * running the JVM proc
+ * 3)  Support pooling/reuse of processes.
+ * 4)  Support any sort of process or session mgmt.  E.g. there's no way to kill or check status of procs invoked
+ *     from the governor.
+ * 5)  Limit threads.  Each .start() request will launch a thread use to acquire a permit, start the proc and
+ *     wait for results.  Consider replacing this design with an Actor pattern.
+ *
+ *
+ * Alternatives considered:
+ *  apache-commons-exec
+ *  https://github.com/zeroturnaround/zt-exec
+ *
+ *  Both have better support for IO handling, but don't pool or limit procs.
+ *
+ *  https://github.com/ViktorC/PP4J
+ *
+ *  Supports pooling, but seems primarily geared at external *java* processes.  Personal project of some guy.
+ */
+public class ProcessGovernorImpl implements ProcessGovernor {
+  private static final Logger logger = LoggerFactory.getLogger( ProcessGovernorImpl.class );
+  private final int maxProcesses;
```

Comment:
While I haven't formulated completely in my mind, it seems like an AtomicInteger could replace the semaphore, synchronized method, and the non thread safe int.

Reply:
Well, the int _is_ threadsafe, by virtue of *final*.

I'd be curious to see what you can come up with using AtomicInteger.  I wouldn't overthink it, though-- if we were to invest much effort here I'd say we should swap to a queue managed by an actor thread, which could get rid of some of the limitations.